File: libsrc/java/org/endurox/TPQCTL.java
/**
 * @brief Enduro/X Queue Control class. Contains details about message
 *   to be enqueued, or information about message dequeued.
 *
 * @class TPQCTL
 */
/* -----------------------------------------------------------------------------
 * Enduro/X Middleware Platform for Distributed Transaction Processing
 * Copyright (C) 2009-2016, ATR Baltic, Ltd. All Rights Reserved.
 * Copyright (C) 2017-2019, Mavimax, Ltd. All Rights Reserved.
 * This software is released under one of the following licenses:
 * LGPL or Mavimax's license for commercial use.
 * See LICENSE file for full text.
 *
 * C (as designed by Dennis Ritchie and later authors) language code is licensed
 * under Enduro/X Modified GNU Affero General Public License, version 3.
 * See LICENSE_C file for full text.
 * -----------------------------------------------------------------------------
 * LGPL license:
 * 
 * This program is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License, version 3 as published
 * by the Free Software Foundation;
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 * PARTICULAR PURPOSE. See the GNU Lesser General Public License, version 3
 * for more details.
 *
 * You should have received a copy of the Lesser General Public License along 
 * with this program; if not, write to the Free Software Foundation, Inc., 
 * 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 *
 * -----------------------------------------------------------------------------
 * A commercial use license is available from Mavimax, Ltd
 * contact@mavimax.com
 * -----------------------------------------------------------------------------
 */
package org.endurox;

import java.util.Arrays;

/**
 * Enduro/X message queue interface.
 * Holds infos about message to be enqueued or dequeued
 */
public class TPQCTL {

    public long getFlags() {
        return flags;
    }

    public void setFlags(long flags) {
        this.flags = flags;
    }

    public long getDeq_time() {
        return deq_time;
    }

    public void setDeq_time(long deq_time) {
        this.deq_time = deq_time;
    }

    public long getPriority() {
        return priority;
    }

    public void setPriority(long priority) {
        this.priority = priority;
    }

    public long getDiagnostic() {
        return diagnostic;
    }

    public void setDiagnostic(long diagnostic) {
        this.diagnostic = diagnostic;
    }

    public String getDiagmsg() {
        return diagmsg;
    }

    public void setDiagmsg(String diagmsg) {
        this.diagmsg = diagmsg;
    }

    public byte[] getMsgid() {
        return msgid;
    }

    public void setMsgid(byte[] msgid) {
        //this.msgid = msgid;
        this.msgid = Arrays.copyOf(msgid, msgid.length);
    }

    public byte[] getCorrid() {
        return corrid;
    }

    public void setCorrid(byte[] corrid) {
        /*
        this.corrid = corrid;
        */
        /* copy off!!! */
        this.corrid = Arrays.copyOf(corrid, corrid.length);
    }

    public String getReplyqueue() {
        return replyqueue;
    }

    public void setReplyqueue(String replyqueue) {
        this.replyqueue = replyqueue;
    }

    public String getFailurequeue() {
        return failurequeue;
    }

    public void setFailurequeue(String failurequeue) {
        this.failurequeue = failurequeue;
    }

    public ClientId getCltid() {
        return cltid;
    }

    public void setCltid(ClientId cltid) {
        this.cltid = cltid;
    }

    public long getUrcode() {
        return urcode;
    }

    public void setUrcode(long urcode) {
        this.urcode = urcode;
    }

    public long getAppkey() {
        return appkey;
    }

    public void setAppkey(long appkey) {
        this.appkey = appkey;
    }

    public long getDelivery_qos() {
        return delivery_qos;
    }

    public void setDelivery_qos(long delivery_qos) {
        this.delivery_qos = delivery_qos;
    }

    public long getReply_qos() {
        return reply_qos;
    }

    public void setReply_qos(long reply_qos) {
        this.reply_qos = reply_qos;
    }

    public long getExp_time() {
        return exp_time;
    }

    public void setExp_time(long exp_time) {
        this.exp_time = exp_time;
    }
	
    /**
     * indicates which of the values are set
     * @HasGetter
     * @HasSetter
     */
    long flags;             /**< indicates which of the values are set */
    long deq_time;          /**< absolute/relative  time for dequeuing */
    long priority;          /**< enqueue priority */
    long diagnostic;        /**< indicates reason for failure */
    String diagmsg;         /**< diagnostic message */
    /** id of message before which to queue */
    byte []msgid = new byte[AtmiConst.TMMSGIDLEN];
    /** correlation id used to identify message */
    byte []corrid = new byte [AtmiConst.TMCORRIDLEN];
    String replyqueue;      /**< queue name for reply message */
    String failurequeue;    /**< queue name for failure message */
    ClientId cltid;         /**< client identifier for originating client */
    long urcode;            /**< application user-return code */
    long appkey;            /**< application authentication client key */
    long delivery_qos;      /**< delivery quality of service  */
    long reply_qos;         /**< reply message quality of service  */
    long exp_time;          /**< expiration time  */
}

/* vim: set ts=4 sw=4 et smartindent: */

File: libsrc/c/TPQCTL.c
/**
 * @brief TPQCTL structure translate from/to Java/C
 *
 * @file TPQCTL.c
 */
/* -----------------------------------------------------------------------------
 * Enduro/X Middleware Platform for Distributed Transaction Processing
 * Copyright (C) 2009-2016, ATR Baltic, Ltd. All Rights Reserved.
 * Copyright (C) 2017-2019, Mavimax, Ltd. All Rights Reserved.
 * This software is released under one of the following licenses:
 * LGPL or Mavimax's license for commercial use.
 * See LICENSE text.
 *
 * C (as designed by Dennis Ritchie and later authors) language code is licensed
 * under Enduro/X Modified GNU Affero General Public License, version 3.
 * See LICENSE_C text.
 * -----------------------------------------------------------------------------
 * LGPL license:
 * 
 * This program is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License, version 3 as published
 * by the Free Software Foundation;
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 * PARTICULAR PURPOSE. See the GNU Lesser General Public License, version 3
 * for more details.
 *
 * You should have received a copy of the Lesser General Public License along 
 * with this program; if not, write to the Free Software Foundation, Inc., 
 * 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 *
 * -----------------------------------------------------------------------------
 * A commercial use license is available from Mavimax, Ltd
 * contact@mavimax.com
 * -----------------------------------------------------------------------------
 */

/*---------------------------Includes-----------------------------------*/
#include <jni.h>
#include <errno.h>
#include <stdlib.h>
#include <exjglue.h>
#include <atmi.h>
#include <oatmi.h>
#include <ndebug.h>
#include <ondebug.h>
#include <oatmisrv_integra.h>
#include "libsrc.h"
#include <sys_unix.h>
/*---------------------------Externs------------------------------------*/
/*---------------------------Macros-------------------------------------*/
#define TPQCTL_CLASS     "org/endurox/TPQCTL"

#define OFSZ(s,e)   EXOFFSET(s,e), EXELEM_SIZE(s,e)

/*---------------------------Enums--------------------------------------*/
/*---------------------------Typedefs-----------------------------------*/
/*---------------------------Globals------------------------------------*/
/*---------------------------Statics------------------------------------*/

/**
 * Only field not mapped is ClientId cltid, this shall be driven manually
 */
exprivate exjobjmap_t M_fieldmap[] =
{    
    {&ndrxj_clazz_TPQCTL_fid_flags,     "flags",       "J",                 OFSZ(TPQCTL,flags)},
    {&ndrxj_clazz_TPQCTL_fid_priority,  "priority",    "J",                 OFSZ(TPQCTL,priority)},
    {&ndrxj_clazz_TPQCTL_fid_diagnostic,"diagnostic",  "J",                 OFSZ(TPQCTL,diagnostic)},
    {&ndrxj_clazz_TPQCTL_fid_diagmsg,   "diagmsg",     "Ljava/lang/String;",OFSZ(TPQCTL,diagmsg)},
    {&ndrxj_clazz_TPQCTL_fid_msgid,     "msgid",       "[B",                OFSZ(TPQCTL,msgid)},
    {&ndrxj_clazz_TPQCTL_fid_corrid,    "corrid",      "[B",                OFSZ(TPQCTL,corrid)},
    {&ndrxj_clazz_TPQCTL_fid_replyqueue, "replyqueue",  "Ljava/lang/String;",OFSZ(TPQCTL,replyqueue)},
    {&ndrxj_clazz_TPQCTL_fid_failurequeue, "failurequeue","Ljava/lang/String;",OFSZ(TPQCTL,failurequeue)},
    {&ndrxj_clazz_TPQCTL_fid_urcode,    "urcode",      "J",                 OFSZ(TPQCTL,urcode)},
    {&ndrxj_clazz_TPQCTL_fid_appkey,    "appkey",      "J",                 OFSZ(TPQCTL,appkey)},
    {&ndrxj_clazz_TPQCTL_fid_delivery_qos, "delivery_qos","J",              OFSZ(TPQCTL,delivery_qos)},
    {&ndrxj_clazz_TPQCTL_fid_reply_qos, "reply_qos",   "J",                 OFSZ(TPQCTL,reply_qos)},
    {&ndrxj_clazz_TPQCTL_fid_exp_time,  "exp_time",    "J",                 OFSZ(TPQCTL,exp_time)},
    {NULL}
};

/*---------------------------Prototypes---------------------------------*/

/**
 * Translate TPQCTL to C from java
 * @param env java env
 * @param ctx_obj ATMI Context
 * @param ctl_Java Java object
 * @param ctl_c C object
 * @return EXSUCCEED/EXFAIL
 */
expublic int ndrxj_atmi_TPQCTL_translate2c(JNIEnv *env, 
            jobject ctx_obj, jobject ctl_Java, TPQCTL *ctl_c)
{
    int ret = EXSUCCEED;
    
    jobject jcltid;
    
    /* Load values to C */
    
    if (EXSUCCEED!=ndrxj_cvt_to_c(env, 
            ctx_obj, M_fieldmap, TPQCTL_CLASS,
            ctl_Java, ctl_c))
    {
        NDRX_LOG(log_error, "Failed to convert %s to TPQCTL!", TPQCTL_CLASS);
        EXFAIL_OUT(ret);
    }
    
    /* get object field */
    jcltid = (*env)->GetObjectField(env, ctl_Java, ndrxj_clazz_TPQCTL_fid_cltid);
    
    if (NULL!=jcltid)
    {
        ndrxj_atmi_throw(env, NULL, NULL, TPEINVAL, "cltid is NULL in TPQCTL!");
        EXFAIL_OUT(ret);

        /* convert to C */
        if (EXSUCCEED!=ndrxj_atmi_ClientId_translate_toc(env, 
            jcltid, &(ctl_c->cltid)))
        {
            NDRX_LOG(log_error, "Failed to convert client id");
            EXFAIL_OUT(ret);
        }
    }
    else
    {
        ctl_c->cltid.clientdata[0] = EXEOS;
    }
    
out:

    if (EXSUCCEED!=ret && !(*env)->ExceptionCheck(env))
    {
        ndrxj_atmi_throw(env, NULL, NULL, TPEINVAL, "Failed to convert TPQCTL to C "
                "from java - see logs!");
    }

    return ret;
}

/**
 * Translate C struct to Java
 * @param env java env
 * @param ctx_obj ATMI Ctx obj
 * @param ctl_Java optional java control struct (use this, if passed)
 * @param ctl_c C Control struct
 * @return EXSUCCEED/EXFAIL
 */
expublic jobject ndrxj_atmi_TPQCTL_translate2java(JNIEnv *env, 
            jobject ctx_obj, jobject ctl_Java, TPQCTL *ctl_c)
{
    int ret = EXSUCCEED;
    jobject retObj = NULL;
    jobject jcltid;

    if (NULL==ctl_Java)
    {
        /* Allocate java object */
        
        NDRX_LOG(log_debug, "About to NewObject(%s)", TPQCTL_CLASS);

        retObj = (*env)->NewObject(env, ndrxj_clazz_TPQCTL, ndrxj_clazz_TPQCTL_mid_INIT);
    }
    else
    {
        NDRX_LOG(log_debug, "Re-use existing java qctl");
        retObj = ctl_Java;
    }
    
    /* Load values to C */
    
    if (EXSUCCEED!=ndrxj_cvt_to_java(env, 
            ctx_obj, M_fieldmap, TPQCTL_CLASS,
            ctl_c, retObj))
    {
        NDRX_LOG(log_error, "Failed to convert C TPQCTL to java %s!", TPQCTL_CLASS);
        EXFAIL_OUT(ret);
    }
    
    /* convert to java */
    if (NULL==(jcltid = ndrxj_atmi_ClientId_translate(env, ctx_obj, EXTRUE,
        &(ctl_c->cltid))))
    {
        NDRX_LOG(log_error, "Failed to convert client id");
        EXFAIL_OUT(ret);
    }
    
    /* set field */
    
    (*env)->SetObjectField(env, retObj, ndrxj_clazz_TPQCTL_fid_cltid, jcltid);
    
out:

    if (NULL!=retObj && EXSUCCEED!=ret && NULL==ctl_Java)
    {
        (*env)->DeleteLocalRef( env, retObj);
    }

    if (EXSUCCEED!=ret && !(*env)->ExceptionCheck(env))
    {
        ndrxj_atmi_throw(env, NULL, NULL, TPEINVAL, "Failed to convert C TPQCTL to java "
                "- see logs!");
    }

    return retObj;
}

/* vim: set ts=4 sw=4 et smartindent: */
